### PR TITLE
Introduce VariableMassInvPC

### PR DESCRIPTION
**Pull request by stephankramer**
Turns out MassInvPC is not actually written for variable (in time) viscosity: it only assembles the preconditioner matrix once and is thus based on the viscosity from the initial condition.

Implement drop-in replacement that does update (in nonlinear iterations and between solves) based on AuxillaryOperatorPC.

Also adds SPDAssembledPC, which is a version of AssembledPC that sets the SPD MatOption. To be used for inner velocity block with gamg. And adds "assembled_pc_gamg_coarse_eq_limit": 1000. These are the 2 "new options" that are recommended for GAMG, but don't seem to affect much in testing.

### CODE REVIEW

File: gadopt/preconditioners.py [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+        appctx = self.get_appctx(pc)
+        V = fd.dmhooks.get_function_space(pc.getDM())
+        # get function spaces
+        assert V.ufl_element().degree() == 0
+        u = fd.TrialFunction(V)
+        v = fd.TestFunction(V)
+        massinv = fd.assemble(fd.Tensor(fd.inner(u, v)*fd.dx).inv)
+        self.massinv = massinv.petscmat
+        self.mu = appctx["mu"]
+        self.gamma = appctx["gamma"]
```
Comment (angus-g) [inline gadopt/preconditioners.py:36]: Should these be `.get()` like in `VariableMassInvPC`, or maybe they'll always be present anyway?

File: gadopt/preconditioners.py [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+    def initialize(self, pc):
+        _, P = pc.getOperators()
+        appctx = self.get_appctx(pc)
+        V = fd.dmhooks.get_function_space(pc.getDM())
+        # get function spaces
+        assert V.ufl_element().degree() == 0
+        u = fd.TrialFunction(V)
+        v = fd.TestFunction(V)
+        massinv = fd.assemble(fd.Tensor(fd.inner(u, v)*fd.dx).inv)
+        self.massinv = massinv.petscmat
```
Comment (angus-g) [inline gadopt/preconditioners.py:34]: Does this get used anywhere? Not that I really know how the Python PC implementations work!

Comment (stephankramer) [reply gadopt/preconditioners.py:34]: Sorry, ignore that P0MassInv. The module comes from an experimental branch, and I thought I'd removed this P0MassInvPC but it seems to have slipped back in. Removing again now....

**Review by angus-g — APPROVED** [review]
Looks good! Will this change convergence (for variable viscosity cases) since the preconditioning changes?

**Comment by stephankramer** [comment]
Probably safer to answer this after the CI has run ;-) but here goes:
It shouldn't (significantly) change the time-stepped convergence to steady state. It should (hopefully significantly) improve the convergence of the solves in each timestep for the variable viscosity cases, yes.

**Comment by angus-g** [comment]
CI took ages to run! Not sure if it was just other stuff running on Kerguelen or what :thinking:

**Comment by angus-g** [comment]
> And adds "assembled_pc_gamg_coarse_eq_limit": 1000. These are the 2 "new options" that are recommended for GAMG, but don't seem to affect much in testing.

Doesn't seem to be doing this, by the way?

Also, runs still taking longer than usual. I can see that it's currently running the `delta/cylindrical` cases. Maybe it's now paying the cost of assembly on all iterations, even without variable viscosity?

**Comment by stephankramer** [comment]
So it turns out that MassInvPC is inconsistent with AssembledPC, which the new VariableMassInvPC is based on: MassInvPC has a ksp directly under it, whereas AssembledPC only has a pc. This means the fieldsplit_1_ksp_Mp_ksp_type cg option is ignored and it currently only does a single apply of SOR. Requires some more refactoring to fix this...

**Comment by angus-g** [comment]
Good sleuthing! That does sound trickier to deal with...

**Comment by angus-g** [comment]
Comparing the timings for master (https://github.com/g-adopt/g-adopt/actions/runs/6436808137/job/17480884421) and this branch (https://github.com/g-adopt/g-adopt/actions/runs/6438893481/job/17485937413), the biggest impact comes from the analytical cases:

| case | master | massinvpc | diff |
| ---- | ------ | --------- | ---- |
| delta/cylindrical/free_slip_dpc 16 8 | 920.630737 | 2525.863525 | **1605.232788** |
| delta/cylindrical/free_slip_dpc 16 2 | 899.034912 | 1681.19458 | **782.159668** |
| smooth/cylindrical/free_slip 16 8 1 | 520.145996 | 1165.673828 | **645.527832** |
| delta/cylindrical/free_slip_dpc 8 8 | 349.433685 | 650.639099 | **301.205414** |

<details><summary>Full table</summary>
<p>

Total positive difference: **3664.546346**
Total negative difference: **-831.439845**
Net difference: **2833.106501**

case | master | massinvpc | diff | ratio
---- | ------ | --------- | ---- | -----
analytical.py run delta/cylindrical/free_slip_dpc 16 8 | 920.630737 | 2525.863525 | 1605.232788 | 2.744
analytical.py run delta/cylindrical/free_slip_dpc 16 2 | 899.034912 | 1681.19458 | 782.159668 | 1.870
analytical.py run smooth/cylindrical/free_slip 16 8 1 | 520.145996 | 1165.673828 | 645.527832 | 2.241
analytical.py run delta/cylindrical/free_slip_dpc 8 8 | 349.433685 | 650.639099 | 301.205414 | 1.862
analytical.py run smooth/cylindrical/free_slip 4 2 4 | 65.367653 | 136.006714 | 70.639061 | 2.081
analytical.py run smooth/cylindrical/zero_slip 16 2 4 | 824.884399 | 880.661072 | 55.776673 | 1.068
analytical.py run smooth/cylindrical/zero_slip 16 8 1 | 842.714783 | 887.020996 | 44.306213 | 1.053
analytical.py run smooth/cylindrical/free_slip 4 2 1 | 106.695328 | 149.239227 | 42.543899 | 1.399
analytical.py run smooth/cylindrical/free_slip 16 8 4 | 491.731079 | 524.567444 | 32.836365 | 1.067
analytical.py run smooth/cylindrical/free_slip 16 2 4 | 585.424133 | 608.463745 | 23.039612 | 1.039
base_case.py | 161.990845 | 173.187225 | 11.19638 | 1.069
compressible_case_ALA.py | 645.213318 | 655.063293 | 9.849975 | 1.015
analytical.py run smooth/cylindrical/free_slip 16 2 1 | 590.889893 | 599.216125 | 8.326232 | 1.014
analytical.py run smooth/cylindrical/free_slip 8 2 4 | 172.995712 | 181.188629 | 8.192917 | 1.047
helmholtz.py | 106.731735 | 112.796974 | 6.065239 | 1.057
analytical.py run smooth/cylindrical/zero_slip 16 2 1 | 833.404053 | 838.63208 | 5.228027 | 1.006
analytical.py run smooth/cylindrical/zero_slip 8 8 4 | 295.257904 | 300.290192 | 5.032288 | 1.017
spiegelman.py 0.0025_1e+23_64_16_50_True | 22.111034 | 24.906487 | 2.795453 | 1.126
spiegelman.py 0.0025_1e+23_64_16_5_False | 14.152982 | 15.852147 | 1.699165 | 1.120
analytical.py run smooth/cylindrical/zero_slip 8 8 1 | 298.042114 | 299.15921 | 1.117096 | 1.004
spiegelman.py 0.0025_1e+23_64_16_0_False | 13.799596 | 14.523582 | 0.723986 | 1.052
analytical.py run delta/cylindrical/zero_slip_dpc 16 2 | 658.879456 | 659.516235 | 0.636779 | 1.001
taylor_test.py Tobs | 1028.620483 | 1029.035767 | 0.415284 | 1.000
analytical.py run delta/cylindrical/zero_slip_dpc 16 8 | 749.220154 | 749.091736 | -0.128418 | 1.000
spiegelman.py 0.0025_1e+23_64_16_50_False | 38.74427 | 38.390518 | -0.353752 | 0.991
taylor_test.py uobs | 1023.897522 | 1022.985352 | -0.91217 | 0.999
compressible_case_TALA.py | 650.341309 | 649.241577 | -1.099732 | 0.998
spiegelman.py 0.0025_1e+23_64_16_15_False | 16.995466 | 15.772203 | -1.223263 | 0.928
3d_spherical.py | 1195.502686 | 1194.249268 | -1.253418 | 0.999
spiegelman.py 0.0025_1e+23_64_16_15_True | 16.627386 | 15.278202 | -1.349184 | 0.919
spiegelman.py 0.0025_1e+23_64_16_5_True | 16.641121 | 14.517623 | -2.123498 | 0.872
analytical.py run delta/cylindrical/zero_slip 8 8 | 486.730988 | 484.593811 | -2.137177 | 0.996
analytical.py run smooth/cylindrical/zero_slip 8 2 1 | 265.351349 | 260.959106 | -4.392243 | 0.983
spiegelman.py 0.0025_1e+23_64_16_0_True | 18.352373 | 13.446807 | -4.905566 | 0.733
forward.py | 141.50563 | 136.225098 | -5.280532 | 0.963
analytical.py run smooth/cylindrical/free_slip 8 8 1 | 189.691833 | 183.53273 | -6.159103 | 0.968
analytical.py run delta/cylindrical/zero_slip_dpc 8 2 | 246.03389 | 237.519577 | -8.514313 | 0.965
analytical.py run smooth/cylindrical/free_slip 8 2 1 | 188.469894 | 179.29274 | -9.177154 | 0.951
analytical.py run smooth/cylindrical/free_slip 8 8 4 | 197.687073 | 184.072327 | -13.614746 | 0.931
analytical.py run delta/cylindrical/free_slip_dpc 8 2 | 298.293182 | 281.20755 | -17.085632 | 0.943
analytical.py run delta/cylindrical/zero_slip 16 8 | 1504.717773 | 1487.316284 | -17.401489 | 0.988
3d_cartesian.py | 2085.177002 | 2067.423828 | -17.753174 | 0.991
analytical.py run smooth/cylindrical/zero_slip 8 2 4 | 294.568573 | 276.295502 | -18.273071 | 0.938
analytical.py run delta/cylindrical/zero_slip_dpc 8 8 | 287.110962 | 268.553009 | -18.557953 | 0.935
analytical.py run delta/cylindrical/free_slip 8 8 | 339.706665 | 313.344818 | -26.361847 | 0.922
analytical.py run delta/cylindrical/free_slip 16 2 | 1003.707581 | 976.757751 | -26.94983 | 0.973
analytical.py run smooth/cylindrical/zero_slip 4 2 1 | 124.711128 | 91.786026 | -32.925102 | 0.736
analytical.py run delta/cylindrical/zero_slip 8 2 | 508.111725 | 470.062744 | -38.048981 | 0.925
analytical.py run delta/cylindrical/free_slip 16 8 | 994.818726 | 950.344482 | -44.474244 | 0.955
analytical.py run delta/cylindrical/free_slip 8 2 | 335.370026 | 285.529663 | -49.840363 | 0.851
analytical.py run smooth/cylindrical/zero_slip 4 2 4 | 137.792938 | 87.384605 | -50.408333 | 0.634
analytical.py run smooth/cylindrical/zero_slip 4 8 1 | 144.068207 | 87.349541 | -56.718666 | 0.606
analytical.py run delta/cylindrical/zero_slip 16 2 | 1518.98584 | 1457.184937 | -61.800903 | 0.959
analytical.py run smooth/cylindrical/zero_slip 4 8 4 | 128.107559 | 64.786957 | -63.320602 | 0.506
analytical.py run smooth/cylindrical/zero_slip 16 8 4 | 849.34845 | 747.012878 | -102.335572 | 0.880
2d_cylindrical.py | 3635.460693 | 3508.900879 | -126.559814 | 0.965

</p>
</details>

**Comment by stephankramer** [comment]
Just rebased on master now to include the lower penalty parameter which (also?) had a big effect on the analytical cases.

**Comment by angus-g** [comment]
The timings above were actually for what is now master, and this branch with master merged into it, so the lower penalty parameter was already included (see https://github.com/g-adopt/g-adopt/commits/angus-g/variable-massinvpc).

**Comment by angus-g** [comment]
Hopefully you're looking at the "Output timing information" step of both those jobs. Perhaps it should be more clear that the finish time in the "Run test" step is the time *since the beginning of the job*, i.e. including queuing time.

I see 56 commands executed on each of the logs you linked, and more or less the same timing trend as the table I posted above: `delta/cylindrical/free_slip_dpc 16 2` gives 886.698975s (master) vs 1930.344360s (massinvpc); `delta/cylindrical/free_slip_dpc 16 8` gives 902.145569s (master) vs 1427.671875s (massinvpc).

**Comment by stephankramer** [comment]
Hm yeah, sorry, you're quite right. I seem to have this bizarre copy and paste issue in my browser - where if I select that output in a certain way, it randomly omits some lines.

Hm actually, now that I have copied and pasted it properly: there are actually 81 timings in that log! Your master.txt also seems to have randomly left out some timings, and they're not actually the same set either as in massinvpc.txt, e.g. there's a "delta/cylindrical/free_slip 4 8" in master.txt but not in massinvpc.txt, and massinvpc.txt has 11 spiegelman cases whereas master.txt only has 9. Seems like copy and pasting from that page is a bit iffy - if that's what you did as well - I guess we should be copy and pasting from the raw logs instead.

In any case, it does seem these larger dpc cases are indeed significantly different still

**Comment by angus-g** [comment]
Oh yes, good catch. How bizarre! :upside_down_face:

**Comment by drhodrid** [comment]
OK so I have been running delta/cylindrical/free_slip_dpc 16 8 on Gadi. I was originally seeing intermittent solver issues, but I now get the following (with solver tolerances updated in line with #31). For G-ADOPT master:

`   Linear Stokes_fieldsplit_0_ solve converged due to CONVERGED_RTOL iterations 44
    Linear Stokes_fieldsplit_0_ solve converged due to CONVERGED_RTOL iterations 43
    Linear Stokes_fieldsplit_0_ solve converged due to CONVERGED_RTOL iterations 42
    Linear Stokes_fieldsplit_0_ solve converged due to CONVERGED_RTOL iterations 43
    Linear Stokes_fieldsplit_0_ solve converged due to CONVERGED_RTOL iterations 45
    Linear Stokes_fieldsplit_0_ solve converged due to CONVERGED_RTOL iterations 44
    Linear Stokes_fieldsplit_1_ solve converged due to CONVERGED_RTOL iterations 5
    Linear Stokes_fieldsplit_0_ solve converged due to CONVERGED_RTOL iterations 43
`
And for VariableMassInvPC:

`
    Linear Stokes_fieldsplit_0_ solve converged due to CONVERGED_RTOL iterations 42
    Linear Stokes_fieldsplit_0_ solve converged due to CONVERGED_RTOL iterations 42
    Linear Stokes_fieldsplit_0_ solve converged due to CONVERGED_RTOL iterations 41
    Linear Stokes_fieldsplit_0_ solve converged due to CONVERGED_RTOL iterations 42
    Linear Stokes_fieldsplit_0_ solve converged due to CONVERGED_RTOL iterations 43
    Linear Stokes_fieldsplit_0_ solve converged due to CONVERGED_RTOL iterations 43
    Linear Stokes_fieldsplit_1_ solve converged due to CONVERGED_RTOL iterations 5
    Linear Stokes_fieldsplit_0_ solve converged due to CONVERGED_RTOL iterations 41
`

Run-times are practically indistinguishable. I note that this is with the build on Mark's fast filter branch, rather than PETSc upstream. For PETSc upstream (which I think is what the CI is testing), I think we need to set the following solver options:

    stokes_solver.solver_parameters['fieldsplit_0']["assembled_pc_gamg_mis_k_minimum_degree_ordering"] = False
    stokes_solver.solver_parameters['fieldsplit_0']["assembled_pc_gamg_aggressive_square_graph"] = True

Do you want to add these to your branch for now (and await merge of my tolerance change branch) to see if that helps?

**Comment by angus-g** [comment]
That seems to have done the trick, they now finish within a couple of minutes. Biggest increase in time is `delta/cylindrical/zero_slip 16 8` at 121s, but on the other hand `2d_cylindrical` takes 156s less, and `3d_spherical` takes 269s less!

**Comment by drhodrid** [comment]
Great that things are now more in line. 

I don't think this necessarily needs to delay merging (perhaps @stephankramer can comment), but I'm generally still trying to understand these timings. For example, the two cases that got quicker (2d_cylindrical and 3d_spherical) are isoviscous - I did not change their tolerances, and the new variable-massinvpc preconditioner should be exactly like the old (because the viscosity is constant in space and time). So why are they quicker?

Is this perhaps a limitation on running these CI cases on a system that is used by others intermittently at different loads? Would these differences in time be expected from that?

**Comment by angus-g** [comment]
I think it's more likely because those two cases run for a very long time (about 1 hour, and 40 minutes, respectively). So the relative change is pretty small. Both CI runs happened overnight where I doubt there was any other load. So it's probably more a case of different scheduling of other cases within the job during those two long cases.

**Comment by drhodrid** [comment]
Ahhh OK. In that case, I think this is good to go. I'm hopnig the tolerance shift also allows the diagonal shift to pass, so that can be next!

**Comment by drhodrid** [comment]
Give me 20 misn to test something @angus-g - and then pending that we should be good to merge.

**Comment by drhodrid** [comment]
So what I've been testing is the differences above in the number of inner solve iterations between master and this branch. 

This variablemassinv pc branch also changes a few default solver parameters (i.e. coarse_eq_limit) and uses the SPD stuff. When adding these options to a case run using master, I get identical results to this branch:

    Linear Stokes_fieldsplit_0_ solve converged due to CONVERGED_RTOL iterations 42
    Linear Stokes_fieldsplit_0_ solve converged due to CONVERGED_RTOL iterations 42
    Linear Stokes_fieldsplit_0_ solve converged due to CONVERGED_RTOL iterations 41
    Linear Stokes_fieldsplit_0_ solve converged due to CONVERGED_RTOL iterations 42
    Linear Stokes_fieldsplit_0_ solve converged due to CONVERGED_RTOL iterations 43
    Linear Stokes_fieldsplit_0_ solve converged due to CONVERGED_RTOL iterations 43
    Linear Stokes_fieldsplit_1_ solve converged due to CONVERGED_RTOL iterations 5
    Linear Stokes_fieldsplit_0_ solve converged due to CONVERGED_RTOL iterations 41

This confirms that Steph's changes do not influence the inner solve in any way. Given timings are now lining up too, we are good to go. Merge away!

**Pull request merged**